README.pod
./lib/Mojolicious/Plugin/FormTamperingProtector.pm

Changes
Revision history for Mojolicious::Plugin::FormTamperingProtector

0.04 2016/12/01
  - Renamed plugin from M::P::FormValidatorLazy.

0.03 2015/12/22
  - Updated dependent version of mojo to v6.38

0.02 2015/02/27
  - Updated internals in favor of Mojolicious v6.0

0.01 2015/02/20
  - initial release


t/test_app.pl
#!/usr/bin/env perl
use FindBin;
use lib "$FindBin::Bin/../lib";
use Mojolicious::Lite;

my $token_key_prefix = 'FormValidatorLazy';

app->secret('afewfweweuhu');

plugin form_validator_lazy => {
    namespace => $token_key_prefix,
    action => '/receptor1',
    blackhole => sub {
        $_[0]->res->code(400);
        $_[0]->render(text => $_[1]);
    },
};

get '/test1' => sub {
    shift->render('test1');
};

post '/receptor1' => sub {
    shift->render(text => 'post completed');
};

post '/receptor2' => sub {
    shift->render(text => 'post completed');
};

app->start;

1;

__END__


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

my %params = (
    NAME          => 'Mojolicious::Plugin::FormValidatorLazy',
    AUTHOR        => 'Sugama Keita <sugama@jamadam.com>',
    VERSION_FROM  => 'lib/Mojolicious/Plugin/FormValidatorLazy.pm',
    ABSTRACT_FROM => 'lib/Mojolicious/Plugin/FormValidatorLazy.pm',
    LICENSE       => 'perl',
    PREREQ_PM     => {
        'Mojolicious' => '6.00',
    },
    BUILD_REQUIRES => {
        'Test::More'            => '0.47',
        'Test::UseAllModules'   => '0.10',
    },
    META_MERGE => {
        resources => {
            repository => {
                type => 'git',
                web  => 'https://github.com/jamadam/Mojolicious-Plugin-FormValidatorLazy',
                url  => 'https://github.com/jamadam/Mojolicious-Plugin-FormValidatorLazy.git',
            },
            bugtracker => {
                web  => 'https://github.com/jamadam/Mojolicious-Plugin-FormValidatorLazy/issues',
            },
        },
    },
    test => {TESTS => 't/*.t t/*/*.t'},
);

my $eumm = $ExtUtils::MakeMaker::VERSION;
delete $params{LICENSE}          if $eumm < 6.31;
delete $params{MIN_PERL_VERSION} if $eumm < 6.48;
delete $params{META_MERGE}       if $eumm < 6.46;
delete $params{META_ADD}         if $eumm < 6.46;
delete $params{LICENSE}          if $eumm < 6.31;

if ($eumm < 6.52 && $params{CONFIGURE_REQUIRES}) {
    $params{PREREQ_PM} = {
        %{ $params{PREREQ_PM}          || {} },
        %{ $params{CONFIGURE_REQUIRES} },
    };
    delete $params{CONFIGURE_REQUIRES};
}
if ($eumm < 6.5503 && $params{BUILD_REQUIRES}) {
    $params{PREREQ_PM} = {
        %{ $params{PREREQ_PM}      || {} },
        %{ $params{BUILD_REQUIRES} },
    };
    delete $params{BUILD_REQUIRES};
}

WriteMakefile(%params);


lib/HTML/ValidationRules/Legacy.pm
package HTML::ValidationRules::Legacy;
use strict;
use warnings;
use Mojo::Base 'Exporter';
use Mojo::JSON;
use Mojo::Util qw{decode};
use Mojo::Parameters;
use Scalar::Util qw(blessed);

our @EXPORT_OK = qw(extract validate),

our $TERM_PROPERTIES         = 'properties';
our $TERM_REQUIRED           = 'required';
our $TERM_MAXLENGTH          = 'maxLength';
our $TERM_MIN_LENGTH         = 'minLength';
our $TERM_OPTIONS            = 'options';
our $TERM_PATTERN            = 'pattern';
our $TERM_MIN                = 'maximam';
our $TERM_MAX                = 'minimum';
our $TERM_TYPE               = 'type';
our $TERM_ADD_PROPS          = 'additionalProperties';
our $TERM_NUMBER             = 'number';

sub extract {
    my ($form, $charset) = @_;
    my $props   = {};
    my @required;
    
    if (! ref $form) {
        $form = Mojo::DOM->new($charset ? decode($charset, $form) : $form);
    }
    
    $form->find("*[name]")->each(sub {
        my $tag = shift;
        my $type = $tag->attr('type') || '';
        my $name = $tag->attr('name');
        $props->{$name} ||= {};
        
        if (grep {$_ eq $type} qw{hidden checkbox radio submit image}) {
            push(@{$props->{$name}->{$TERM_OPTIONS}}, $tag->attr('value'));
        }
        
        if ($tag->type eq 'select') {
            $tag->find('option')->each(sub {
                push(@{$props->{$name}->{$TERM_OPTIONS}}, shift->attr('value'));
            });
        }
        
        if ($type eq 'number') {
            $props->{$name}->{$TERM_TYPE} = $TERM_NUMBER;
            if (my $val = $tag->attr->{min}) {
                $props->{$name}->{$TERM_MIN} = $val;
            }
            if (my $val = $tag->attr->{max}) {
                $props->{$name}->{$TERM_MAX} = $val;
            }
        }
        
        if (! exists $tag->attr->{disabled}) {
            if ($type ne 'submit' && $type ne 'image' && $type ne 'checkbox' &&
                        ($type ne 'radio' || exists $tag->attr->{checked})) {
                $props->{$name}->{$TERM_REQUIRED} = Mojo::JSON->true;
            }
        }
            
        if (exists $tag->attr->{maxlength}) {
            $props->{$name}->{$TERM_MAXLENGTH} = $tag->attr->{maxlength} || 0;
        }
        
        if (exists $tag->attr->{required}) {
            $props->{$name}->{$TERM_MIN_LENGTH} = 1;
        }
        
        if (exists $tag->attr->{pattern}) {
            $props->{$name}->{$TERM_PATTERN} = $tag->attr->{pattern};
        }
    });
    
    return {
        $TERM_PROPERTIES => $props,
        $TERM_ADD_PROPS => Mojo::JSON->false,
    };
}

sub validate {
    my ($schema, $params, $charset) = @_;
    
    if (! (blessed($params) && $params->isa('Mojo::Parameters'))) {
        my $wrapper = Mojo::Parameters->new;
        $wrapper->charset($charset);
        if (blessed($params) && $params->isa('Hash::MultiValue')) {
            $wrapper->append($params->flatten);
        } else {
            $wrapper->append($params);
        }
        $params = $wrapper;
    }
    
    my $props = $schema->{$TERM_PROPERTIES};
    
    if (! $schema->{$TERM_ADD_PROPS}) {
        for my $name ($params->param) {
            return "Field $name is injected" if (! $props->{$name});
        }
    }
    
    for my $name (keys %$props) {
        my @params = grep {defined $_} $params->param($name);
        
        if (($props->{$name}->{$TERM_REQUIRED} || '') eq Mojo::JSON->true) {
            return "Field $name is required" if (! scalar @params);
        }
        
        if (my $allowed = $props->{$name}->{$TERM_OPTIONS}) {
            for my $given (@params) {
                return "Field $name has been tampered"
                                        if (! grep {$_ eq $given} @$allowed);
            }
        }
        if (exists $props->{$name}->{$TERM_MAXLENGTH}) {
            for my $given (@params) {
                return "Field $name is too long"
                    if (length($given) > $props->{$name}->{$TERM_MAXLENGTH});
            }
        }
        if (defined $props->{$name}->{$TERM_MIN_LENGTH}) {
            for my $given (@params) {
                return "Field $name cannot be empty"
                    if (length($given) < $props->{$name}->{$TERM_MIN_LENGTH});
            }
        }
        if (my $pattern = $props->{$name}->{$TERM_PATTERN}) {
            for my $given (@params) {
                return "Field $name not match pattern"
                                                if ($given !~ /\A$pattern\Z/);
            }
        }
        if (($props->{$name}->{$TERM_TYPE} || '') eq $TERM_NUMBER) {
            for my $given (@params) {
                return "Field $name not match pattern"
                                if ($given !~ /\A[\d\+\-\.]+\Z/);
                if (my $min = $props->{$name}->{$TERM_MIN}) {
                    return "Field $name too low" if ($given < $min);
                }
                if (my $max = $props->{$name}->{$TERM_MAX}) {
                    return "Field $name too great" if ($given > $max);
                }
            }
        }
    }
    return; 
}

1;

__END__

=head1 NAME

HTML::ValidationRules::Legacy - Extract Validation Rules from HTML Form Element

=head1 SYNOPSIS

=head1 DESCRIPTION

B<This software is considered to be alpha quality and isn't recommended for
regular usage.>

=head2 CLASS METHODS

=head3 extract

    my $schema = extract($form_in_strig, $charset)
    my $schema = extract($form_in_mojo_dom)

Generates a schema out of form string or Mojo::DOM instance. It returns
schema in hashref consists of JSON-schema-like properties.

=head3 validate

Validates form data against given schema.

    my $error = validate($params_in_string, $schema, $charset);
    my $error = validate($params_in_mojo_params, $schema);

=head1 AUTHOR

Sugama Keita, E<lt>sugama@jamadam.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) Sugama Keita.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


lib/Mojolicious/Plugin/FormTamperingProtecter.pm
package Mojolicious::Plugin::FormTamperingProtecter;
use strict;
use warnings;
use Mojo::Base 'Mojolicious::Plugin';
our $VERSION = '0.01';
use Data::Dumper;
use Mojo::JSON;
use Mojo::Util qw{encode xml_escape hmac_sha1_sum secure_compare};

    has 'secret';
    has 'prefix';
    
    my $DIGEST_INDEX_OPTIONS    = 0;
    my $DIGEST_INDEX_ALLOW_NULL = 1;
    my $DIGEST_INDEX_MAXLENGTH  = 2;
    my $DIGEST_INDEX_REQUIRED   = 3;
    
    my $json = Mojo::JSON->new;
    
    ### ---
    ### register
    ### ---
    sub register {
        my ($self, $app, $options) = @_;
        
        $self->secret($app->secret);
        $self->prefix($options->{token_key_prefix});
        
        $app->hook('around_dispatch' => sub {
            (my $app, my $c) = @_;
            
            my @actions =
                ref $options->{action} ? @{$options->{action}} : $options->{action};
            
            if ($c->req->method eq 'POST' && grep {$_ eq $c->req->url->path} @actions) {
                if (my $error = $self->tampered($c, $self->prefix)) {
                    return $options->{blackhole}->($c, $error);
                }
            }
            
            $app->();
            
            my $dom = $c->res->dom;
            
            for my $action (@actions) {
                $dom->find(qq{form[action="$action"][method="post"]})->each(sub {
                    $self->inject_token(shift, $self->prefix);
                });
            }
            
            $c->res->body(encode('UTF-8', $dom));
        });
    }
    
    sub inject_token {
        my ($self, $form, $prefix) = @_;
        my $names = {};
        $form->find("*:not([disabled])[name]")->each(sub {
            my $tag = shift;
            my $type = $tag->attrs('type');
            my $name = $tag->attrs('name');
            $names->{$name} ||= [];
            if (grep {$_ eq $type} qw{hidden checkbox radio}) {
                push(@{$names->{$name}->[$DIGEST_INDEX_OPTIONS]}, $tag->attrs('value'));
            }
            if (grep {$_ eq $type} qw{hidden}) {
                $names->{$name}->[$DIGEST_INDEX_ALLOW_NULL] = 0;
            }
            if (grep {$_ eq $type} qw{checkbox radio}) {
                $names->{$name}->[$DIGEST_INDEX_ALLOW_NULL] //= 1;
            }
            my $maxlength = $tag->attrs('maxlength');
            if ($maxlength =~ /./) {
                $names->{$name}->[$DIGEST_INDEX_MAXLENGTH] = $maxlength;
            }
            if (grep {$_ eq 'required'} keys %{$tag->attrs}) {
                $names->{$name}->[$DIGEST_INDEX_REQUIRED] = 1;
            }
        });
        
        my $digest = sign($json->encode($names), $self->secret);
        
        $form->append_content(
            sprintf(qq!<input type="hidden" name="%s-token" value="%s">!,
                    $prefix, xml_escape $digest));
    }
    
    sub tampered {
        my ($self, $c, $prefix) = @_;
        
        my $token = $c->param("$prefix-token");

        if (! $token) {
            return 'Token is not found';
        }
        
        my $unsigned = unsign($token, $self->secret);
        
        if (! $unsigned) {
            return 'Token has been tampered';
        }
        
        my $digest = $json->decode($unsigned);
        my @form_names = grep {$_ ne "$prefix-token"} $c->param;
        
        for my $name (@form_names) {
            if (! $digest->{$name}) {
                return "Field $name is injected";
            }
        }
        for my $name (keys %{$digest}) {
            if (! grep {$_ eq $name} @form_names) {
                if (! $digest->{$name}->[$DIGEST_INDEX_ALLOW_NULL]) {
                    return "Field $name is not given";
                }
            }
            if (my $allowed = $digest->{$name}->[$DIGEST_INDEX_OPTIONS]) {
                my $given = scalar $c->param($name);
                if (defined $given && ! grep {$_ eq $given} @$allowed) {
                    return "Field $name has been tampered";
                }
            }
            my $maxlength = $digest->{$name}->[$DIGEST_INDEX_MAXLENGTH];
            if (defined $maxlength) {
                my $given_length = length(scalar $c->param($name));
                if ($given_length > $maxlength) {
                    return "Field $name is too long";
                }
            }
            if (defined $digest->{$name}->[$DIGEST_INDEX_REQUIRED]) {
                my $given = scalar $c->param($name);
                if (! $given || length($given) == 0) {
                    return "Field $name cannot be empty";
                }
            }
        }
        return;
    }
    
    sub sign {
        my ($value, $secret) = @_;
        return $value. '--' . hmac_sha1_sum($value, $secret);
    }
    
    sub unsign {
        my ($signed, $secret) = @_;
        if ($signed =~ s/--([^\-]+)$//) {
            my $sig = $1;
            if (secure_compare($sig, hmac_sha1_sum($signed, $secret))) {
                return $signed;
            }
        }
    }

1;

__END__

=head1 NAME

Mojolicious::Plugin::FormTamperingProtecter - FormTamperingProtecter

=head1 SYNOPSIS
  
=head1 DESCRIPTION

=head2 OPTIONS

=head1 AUTHOR

Sugama Keita, E<lt>sugama@jamadam.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Sugama Keita.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


lib/Mojolicious/Plugin/FormValidatorLazy.pm
package Mojolicious::Plugin::FormValidatorLazy;
use strict;
use warnings;
use Mojo::Base 'Mojolicious::Plugin';
our $VERSION = '0.02';
use Data::Dumper;
use Mojo::JSON qw(decode_json encode_json);
use Mojo::Util qw{encode decode xml_escape hmac_sha1_sum secure_compare
                                                        b64_decode b64_encode};
use HTML::ValidationRules::Legacy qw{validate extract};

our $TERM_ACTION = 0;
our $TERM_SCHEMA = 1;

### ---
### register
### ---
sub register {
    my ($self, $app, $opt) = @_;
    
    my $schema_key = $opt->{namespace}. "-schema";
    my $sess_key = $opt->{namespace}. '-sessid';
    
    my $actions = ref $opt->{action} ? $opt->{action} : [$opt->{action}];
    
    $app->hook(before_dispatch => sub {
        my $c = shift;
        my $req = $c->req;
        
        if ($req->method eq 'POST' && grep {$_ eq $req->url->path} @$actions) {
            
            my $wrapper = deserialize(unsign(
                $req->param($schema_key),
                ($c->session($sess_key) || ''). $app->secrets->[0]
            ));
            
            $req->params->remove($schema_key);
            
            if (!$wrapper) {
                return $opt->{blackhole}->($c,
                            'Form schema is missing, possible hacking attempt');
            }
            if ($req->url->path ne $wrapper->{$TERM_ACTION}) {
                return $opt->{blackhole}->($c,
                                        'Action attribute has been tampered');
            }
            
            if (my $err = validate($wrapper->{$TERM_SCHEMA}, $req->params)) {
                return $opt->{blackhole}->($c, $err);
            }
        }
    });
    
    $app->hook(after_dispatch => sub {
        my $c = shift;
        
        if ($c->res->headers->content_type =~ qr{^text/html} &&
                                                $c->res->body =~ qr{<form\b}i) {
            
            my $sessid = $c->session($sess_key);
            
            if (! $sessid) {
                $sessid = hmac_sha1_sum(time(). {}. rand(), $$);
                $c->session($sess_key => $sessid);
            }
            
            $c->res->body(inject(
                $c->res->body,
                $actions,
                $schema_key,
                $sessid. $app->secrets->[0],
                $c->res->content->charset)
            );
        }
    });
}

sub inject {
    my ($html, $actions, $token_key, $secret, $charset) = @_;
    
    if (! ref $html) {
        $html = Mojo::DOM->new($charset ? decode($charset, $html) : $html);
    }

    $html->find(qq{form[action][method="post"]})->each(sub {
        my $form    = shift;
        my $action  = $form->attr('action');
        
        return if (! grep {$_ eq $action} @$actions);
        
        my $wrapper = sign(serialize({
            $TERM_ACTION    => $action,
            $TERM_SCHEMA    => extract($form, $charset),
        }), $secret);
        
        $form->append_content(sprintf(<<"EOF", $token_key, xml_escape $wrapper));
<div style="display:none">
    <input type="hidden" name="%s" value="%s">
</div>
EOF
    });
    
    return encode($charset, $html);
}

sub serialize {
    return b64_encode(encode_json(shift // return), '');
}

sub deserialize {
    return decode_json(b64_decode(shift // return));
}

sub sign {
    my ($value, $secret) = @_;
    return $value. '--' . hmac_sha1_sum($value, $secret);
}

sub unsign {
    my ($value, $secret) = @_;
    if ($value && $secret && $value =~ s/--([^\-]+)$//) {
        my $sig = $1;
        return $value if (secure_compare($sig, hmac_sha1_sum($value, $secret)));
    }
    return;
}

1;

__END__

=head1 NAME

Mojolicious::Plugin::FormValidatorLazy - FormValidatorLazy

=head1 SYNOPSIS

    plugin form_validator_lazy => {
        namespace => 'form_validator_lazy',
        action => ['/receptor1'],
        blackhole => sub {
            my ($c, $error) = @_;
            app->log($error);
            $c->res->code(400);
            $c->render(text => 'An error occured');
        },
    };

=head1 DESCRIPTION

B<This software is considered to be alpha quality and isn't recommended for
regular usage.>

Mojolicious::Plugin::FormValidatorLazy is a Mojolicious plugin for validating
post data with auto-generated validation rules out of original forms.
It analizes the HTML forms before sending them to client, generate the schema,
inject it into original forms within a hidden fields so the plugin can detect
the schema when a post request comes.

The plugin detects following error for now.

=over

=item Unknown form fields.

The form fields represented by name attribute are all white listed and post data
injected unknown fields are blocked.

=item Unknown values of selectable fields.

Selectable values of checkboxes, radio buttons and select options are white
listed and unknow values are blocked.

The plugin also detects characteristics of tag types. Such as unchecked
checkboxes don't appear to data(not required), radio buttons can't be null only
when default value is offered(not null), and so on.

=item Hidden field tamperings.

Hidden typed input can't be ommited(required) and the value takes only one
option. the plugin blocks values against the schema.

=item Values against maxlength attributes.

Values violating of maxlength are blocked.

=item HTML5 validation attributes

HTML5 supports some validation attributes such as [required], [pattern=*],
[type=number], [min=*], [max=*]. The plugin detects them and block violations.

=item CSRF

This also detects CSRF.

=back

=head2 EXAMPLE

Run t/test_app.pl and try to attack the forms.

    ./t/test_app.pl daemon

=head2 CLASS METHODS

=head3 inject

Generates a schema strings of form structure for each forms in mojo response
and inject them into itself.

    my $injected = inject($html, $charset,
                                ['/path1', '/path2'], $token_key, $session_id);

=head1 AUTHOR

Sugama Keita, E<lt>sugama@jamadam.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) Sugama Keita.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut
